libkbfs/bare_root_metadata_v3.go

```diff
@@ -0,0 +1,970 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-codec/codec"
+)
+
+// WriterMetadataV3 stores the metadata for a TLF that is
+// only editable by users with writer permissions.
+type WriterMetadataV3 struct {
+	// Serialized, possibly encrypted, version of the PrivateMetadata
+	SerializedPrivateMetadata []byte `codec:"data"`
+
+	// The last KB user with writer permissions to this TLF
+	// who modified this WriterMetadata
+	LastModifyingWriter keybase1.UID `codec:"lmw"`
+
+	// For public TLFs (since those don't have any keys at all).
+	Writers []keybase1.UID `codec:",omitempty"`
+	// Writers identified by unresolved social assertions.
+	UnresolvedWriters []keybase1.SocialAssertion `codec:"uw,omitempty"`
+	// Pointer to the writer key bundle for private TLFs.
+	WKeyBundleID TLFWriterKeyBundleID `codec:"wkid,omitempty"`
+	// Latest key generation.
+	LatestKeyGen KeyGen `codec:"lkg"`
+
+	// The directory ID, signed over to make verification easier
+	ID TlfID
+	// The branch ID, currently only set if this is in unmerged per-device history.
+	BID BranchID
+	// Flags
+	WFlags WriterFlags
+
+	// Estimated disk usage at this revision
+	DiskUsage uint64
+	// The total number of bytes in new blocks
+	RefBytes uint64
+	// The total number of bytes in unreferenced blocks
+	UnrefBytes uint64
+
+	codec.UnknownFieldSetHandler
+}
+
+// BareRootMetadataV3 is the MD that is signed by the reader or
+// writer. Unlike RootMetadata, it contains exactly the serializable
+// metadata.
+type BareRootMetadataV3 struct {
+	// The metadata that is only editable by the writer.
+	WriterMetadata WriterMetadataV3 `codec:"wmd"`
+	// The signature for the writer metadata, to prove
+	// that it's only been changed by writers.
+	WriterMetadataSigInfo SignatureInfo `codec:"wmdsi"`
```

Comment:
Is it cryptographically necessary for `WriterMetadataSigInfo` to be in BRMDV3? If it can be put in RootMetadataSigned instead, that would work out better for journaling, since one wouldn't have to compute a signature (relatively expensive) to compute the MD ID for a stored BRMDV3, and can delay that until it's being sent to the server.

That would mean we'd have to version RMDSigned, too, though. Sigh.